odoo/tools/profiler.py

```diff
@@ -1,197 +1,401 @@
+        self.params = params or {}  # custom parameters usable by collectors
+
+        if db is ...:
+            # determine database from current thread
+            db = getattr(threading.current_thread(), 'dbname', None)
+            if not db and not path:
+                # only raise if path is not given and db is not explicitely disabled
+                raise Exception('Database name cannot be defined automaticaly. \n Please provide a valid/falsy dbname or path parameter')
+        self.db = db
+
+        # collectors
+        if collectors is None:
+            collectors = ['sql', 'traces_async']
+        self.collectors = []
+        for collector in collectors:
+            if isinstance(collector, str):
+                try:
+                    collector = Collector.make(collector)
+                except Exception:
+                    _logger.error("Could not create collector with name %r", collector)
                     continue
-                next_call = v['calls'][n]
-                if next_call['queries'] is not None:
-                    v['report'][call['lineno']]['nb_queries'] += next_call['queries'] - call.get('queries', 0)
-                v['report'][call['lineno']]['delay'] += next_call['time'] - call['time']
-
-            queries = 0
-            delay = 0
-            for call in v['report'].values():
-                queries += call['nb_queries']
-                delay += call['delay']
-
-            if minimum_time and minimum_time > delay*1000:
-                continue
-            if minimum_queries and minimum_queries > queries:
-                continue
+            collector.profiler = self
+            self.collectors.append(collector)
+
+    def __enter__(self):
+        self.init_thread = threading.current_thread()
+        self.init_frame = get_current_frame(self.init_thread)
+        if self.init_frame.f_code.co_name == 'enter_context':
+            self.init_frame = self.init_frame.f_back  # profiler used in a ExitStack case
+        self.init_stack_trace = _get_stack_trace(self.init_frame)
+        if self.description is None:
+            frame = self.init_frame
+            code = frame.f_code
+            self.description = f"{frame.f_code.co_name} ({code.co_filename}:{frame.f_lineno})"
+        if self.disable_gc and gc.isenabled():
+            gc.disable()
+        self.start_time = time.time()
+        for collector in self.collectors:
+            collector.start()
+        return self
+
+    def __exit__(self, *args):
+        try:
+            for collector in self.collectors:
+                collector.stop()
+            self.duration = time.time() - self.start_time
+            self._add_file_lines(self.init_stack_trace)
+            if self.path:
+                save_time = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+                dirname = os.path.dirname(self.path)
+                if not os.path.exists(dirname):
+                    os.makedirs(dirname)
```

Comment:
```suggestion
                    os.makedirs(dirname, exists_ok=True)
```
might be a good idea.